examples/conform.cpp

```diff
@@ -0,0 +1,163 @@
+    // new_media = shot->latest_render("mov");
+    
+    const auto matches = glob(folder, name + ".*");
+    
+    if (matches.size() == 0)
+    {
+        //std::cout << "DEBUG: No match for clip '" << name << "'" << std::endl;
+        return std::string();
+    }
+    if (matches.size() == 1)
+    {
+        return matches[0];
+    }
+    else
+    {
+        std::cout << "WARNING: " << matches.size() << " matches found for clip '" <<
+            name << "', using '" << matches[0] << "'";
+        return matches[0];
+    }
+}
+
+// Look for replacement media for each clip in the given timeline.
+//
+// The clips are relinked in place if media with a matching name is found.
+int conform_timeline(
+    otio::SerializableObject::Retainer<otio::Timeline> const& timeline,
+    std::string const& folder)
+{
+    int count = 0;
+    
+    otio::ErrorStatus error_status;
+    const auto clips = timeline->clip_if(&error_status);
+    if (error_status)
+    {
+        std::cout << error_status.outcome << std::endl;
+        throw error_status;
+    }
+    
+    for (auto clip : clips)
+    {
+        // look for a media file that matches the clip's name
+        const auto new_path = find_matching_media(clip->name(), folder);
+
+        // if no media is found, keep going
+        if (new_path.empty())
+            continue;
+
+        // if we found one, then relink to the new path
+        clip->set_media_reference(new otio::ExternalReference(
+            "file://" + new_path,
+            otio::nullopt // we don't know the available range
+        ));
+        count += 1;
+    }
+    
+    return count;
+}
+
+int main(int argc, char** argv)
+{
+    if (argc != 4)
+    {
+        std::cout << "Usage: conform (input) (folder) (output)" << std::endl;
+        return 1;
+    }
+    const std::string input = normalize_path(argv[1]);
+    const std::string folder = normalize_path(argv[2]);
+    const std::string output = normalize_path(argv[3]);
+    
+    try
+    {   
+        otio::ErrorStatus error_status;
+        otio::SerializableObject::Retainer<otio::Timeline> timeline(
+            dynamic_cast<otio::Timeline*>(otio::Timeline::from_json_file(input, &error_status)));
+        if (!timeline || error_status)
+        {
+            throw error_status;
+        }
+        const int count = conform_timeline(timeline, folder);
+        std::cout << "Relinked " << count << " clips to new media." << std::endl;
+        if (!timeline.value->to_json_file(output, &error_status))
+        {
+            throw error_status;
+        }
+        const auto clips = timeline->clip_if(&error_status);
+        if (error_status)
+        {
+            throw error_status;
+        }
+        std::cout << "Saved " << output << " with " << clips.size() << " clips." << std::endl;
+    }
+    catch (otio::ErrorStatus const& e)
+    {
+        print_error(e);
+        return 1;
+    }
+    
+    return 0;
+}
+
```

Comment:
```suggestion
    return 0;
}
catch (std::exception& exc)
{
    std::cerr << "Unexpected termination due to " << exc.what() << std::endl;
    return 1;
}
```

Reply:
Interesting, I haven't see a try/catch block used like that with main() before. Maybe exit() would be preferable since exceptions aren't used by OTIO?

Reply:
exit(1) is fine/preferable, I'd say. Pedantic aside: This try on main pattern is semantically equivalent to having a try/catch inside main. It's more useful on object constructors, like Foo() try { throw exc(); } catch(exc&) { printf("wut"); } ~~~ weirdly the catch after constructors and destructors automatically rethrows, but it's a chance to do something special after an exception.

Reply:
On another note I find some of the error handling a bit confusing with the C++ API; for example ```otio::Timeline::from_json_file()``` can return an error as either a null timeline or an error status. But could it return a non-null timeline and an error status? If so is the timeline  OK to use and the error is recoverable?

Similar also with ```otio::Timeline::to_json_file()```, it returns a bool and an error status. Could it ever return true but with a non-OK error status?

Reply:
The way it's coded, if an error occurred, nullptr is returned. No guarantees are made about ErrorStatus. Since you passed in the original ErrorStatus, it might already have a value; it's not reset by OTIO. So the reasonable check is on the return value, with the expectation that additional new information is available in the error status if the return indicated an error. I notice there is unfortunately no documentation in the header file on these functions.